### PR TITLE
Enable keras estimator to accept modelCheckPoint callback

### PR DESCRIPTION
**Pull request by abditag2**
## Checklist before submitting

- [ ] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

File: horovod/spark/keras/estimator.py [hunk 1]
```diff
@@ -160,6 +160,8 @@ class KerasEstimator(HorovodEstimator, KerasEstimatorParamsReadable,
 
     custom_objects = Param(Params._dummy(), 'custom_objects', 'custom objects')
     _keras_pkg_type = Param(Params._dummy(), '_keras_pkg_type', 'keras package type')
+    model_checkpoint_callback_fn = Param(Params._dummy(), 'model_checkpoint_callback_fn',
```
Comment (tgaddair) [inline horovod/spark/keras/estimator.py:163]: I think `checkpoint_callback` would be a better name, to be consistent with PyTorch Lightning.

File: horovod/spark/keras/remote.py [hunk 2]
```diff
@@ -323,3 +329,10 @@ def fn(tf, keras):
         config.intra_op_parallelism_threads = 1
         keras.backend.set_session(tf.Session(config=config))
     return fn
+
+
+def _model_checkpoint_callback_fn():
```
Comment (tgaddair) [inline horovod/spark/keras/remote.py:334]: I think we can simplify this a bit:

1. Change it from accepting a function to just accepting something that adheres to the `ModelCheckpoint` API.
2. Use the `checkpoint_callback.filepath` property to figure out where to read the saved model from (note that this filepath might be a template string: https://www.tensorflow.org/tutorials/keras/save_and_load#checkpoint_callback_options).
3. If not specified, create the default `ModelCheckpoint(ckpt_file)`.

Comment (abditag2) [reply horovod/spark/keras/remote.py:334]: I don't think that the filepath should come from the user because they have to be responsible to figure out the appropriate temp folder with correct permissions. It also increases the chances of errors if users specify a fixed path, and the checkpoints from previous runs are present on that path. If a user wants copies of the checkpoint, they can add extra ModelCheckpoint callbacks using the callbacks parameter. 

If we agree that the user should not pass the filepath arg, I am not sure how to avoid not returning a function here as we want the user to specify the parameters of ModelCheckpoint callback.

Comment (tgaddair) [reply horovod/spark/keras/remote.py]: There are many cases where the user will want to keep their checkpoints in a particular location.  Needing to provide two separate callbacks that do the same thing in different directories is a bad user experience.

If you want to support the user not having to specify the path (because they don't want to keep the checkpoint files around), then you can solve this by creating a subclass of `ModelCheckpoint` for this purpose:

```
KerasEstimator(checkpoint_callback=hvd.callbacks.ReturnedModelCheckpoint(
    monitor='val_acc',
    mode='max',
    save_best_only=True), ...)
```

Nice thing about this approach is (1) it's flexible if the user wants to provide an alternative location to write checkpoints, and (2) you can prevent them from setting unnecessary things like `filepath` and `save_weights_only`.

File: horovod/keras/callbacks.py [hunk 3]
```diff
@@ -152,3 +152,24 @@ def __init__(self, warmup_epochs=5, momentum_correction=True, steps_per_epoch=No
         """
         super(LearningRateWarmupCallback, self).__init__(K, warmup_epochs, momentum_correction,
                                                          steps_per_epoch, verbose, initial_lr)
+
+
+class ReturnedModelCheckpoint(keras.callbacks.ModelCheckpoint):
+    def __init__(self, filepath=None,
+                 monitor='val_loss',
+                 verbose=0,
+                 save_best_only=True,
```
Comment (tgaddair) [inline horovod/keras/callbacks.py:161]: Is `save_best_only` needed here?  It seems that with a specific file path provided, it always saves the best only.

File: horovod/keras/callbacks.py [hunk 4]
```diff
@@ -152,3 +152,24 @@ def __init__(self, warmup_epochs=5, momentum_correction=True, steps_per_epoch=No
         """
         super(LearningRateWarmupCallback, self).__init__(K, warmup_epochs, momentum_correction,
                                                          steps_per_epoch, verbose, initial_lr)
+
+
+class ReturnedModelCheckpoint(keras.callbacks.ModelCheckpoint):
+    def __init__(self, filepath=None,
```
Comment (tgaddair) [inline horovod/keras/callbacks.py:158]: I think we should not let the user specify `filepath`.  If they want a custom directory, they can just use the regular `ReturnedModelCheckpoint`.  Otherwise, there is no practical difference between this and `ModelCheckpoint`.

File: horovod/keras/callbacks.py (hunk 3 above)
Comment (abditag2) [reply horovod/keras/callbacks.py]: Without `save_best_only`, in every epoch or whatever the monitoring interval is it overwrites the previous checkpoint. With it, it will only overwrite if the current checkpoint has better metric than the previously saved checkpoint. 

I think we need it.

File: horovod/keras/callbacks.py (hunk 4 above)
Comment (abditag2) [reply horovod/keras/callbacks.py:158]: sure

File: horovod/keras/callbacks.py (hunk 3 above)
Comment (tgaddair) [reply horovod/keras/callbacks.py]: Makes sense, but is there ever a case where the user would want to set `save_best_only=False` using this class?

Comment (abditag2) [reply horovod/keras/callbacks.py]: I cannot think of a use case off the top of my head. But I think, it is a better choice to conform with the defaults of the `ModelCheckpoint` args to avoid confusion for the users.

Comment (abditag2) [reply horovod/keras/callbacks.py]: I am open to change this if you have a strong preference.

Comment (tgaddair) [reply horovod/keras/callbacks.py]: My concern is that if we expose `save_best_only` to the user, but it doesn't do anything, that would create even more confusion for the user.  I think for now we may want to remove it, and always set `save_best_only=True` when initializing the parent constructor.
